Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -108,6 +97,52 @@ public struct BSONDocumentIterator: IteratorProtocol {
             }
         }
     }
+
+    // uses an iterator to copy (key, value) pairs of the provided document from range [startIndex, endIndex) into a new
+    // document. starts at the startIndex-th pair and ends at the end of the document or the (endIndex-1)th index,
+    // whichever comes first.
+    internal static func subsequence(
+        of doc: BSONDocument,
+        startIndex: Int = 0,
+        endIndex: Int = Int.max
+    ) -> BSONDocument {
+        // TODO(SWIFT-911): Improve performance
+        guard endIndex >= startIndex else {
+            fatalError("endIndex must be >= startIndex")
+        }
+
+        var iter = BSONDocumentIterator(over: doc)
+
+        var excludedKeys: [String] = []
+
+        for _ in 0..<startIndex {
+            guard let next = iter.next() else {
+                // we ran out of values
+                break
+            }
+            excludedKeys.append(next.key)
+        }
+
+        // skip the values between startIndex and endIndex. this has better performance than calling next, because
+        // it doesn't pull the unneeded key/values out of the iterator
+        for _ in startIndex..<endIndex {
+            if (try? iter.nextThrowing()) == nil {
```

Comment:
This, and really any conditional whose body unconditionally leads to a scope exit, should be written as a `guard`. Doing so allows for easy detection of invariants by just looking for the keyword, and because the compiler enforces that a `guard` must exit scope in its body, it decreases the complexity of the following guarded section. It also of course eliminates the possibility of forgetting a `break` / `return` / `throw`.